src/Authentication/Authenticators/JWT.php

```diff
@@ -0,0 +1,240 @@
+<?php
+
+namespace CodeIgniter\Shield\Authentication\Authenticators;
+
+use CodeIgniter\HTTP\IncomingRequest;
+use CodeIgniter\I18n\Time;
+use CodeIgniter\Shield\Authentication\AuthenticationException;
+use CodeIgniter\Shield\Authentication\AuthenticatorInterface;
+use CodeIgniter\Shield\Authentication\Authenticators\JWT\Firebase;
+use CodeIgniter\Shield\Authentication\Authenticators\JWT\JWTDecoderInterface;
+use CodeIgniter\Shield\Entities\User;
+use CodeIgniter\Shield\Exceptions\RuntimeException;
+use CodeIgniter\Shield\Models\TokenLoginModel;
+use CodeIgniter\Shield\Models\UserModel;
+use CodeIgniter\Shield\Result;
+use InvalidArgumentException;
+use stdClass;
+
+/**
+ * Stateless JWT Authenticator
+ */
+class JWT implements AuthenticatorInterface
+{
+    /**
+     * @var string Special ID Type.
+     *             This Authenticator is stateless, so no `auth_identities` record.
+     */
+    public const ID_TYPE_JWT = 'jwt';
+
+    /**
+     * The persistence engine
+     */
+    protected UserModel $provider;
+
+    protected ?User $user = null;
+    protected JWTDecoderInterface $jwtDecoder;
+    protected TokenLoginModel $loginModel;
+    protected ?stdClass $payload = null;
+
+    public function __construct(UserModel $provider, ?JWTDecoderInterface $jwtDecoder = null)
+    {
+        $this->provider   = $provider;
+        $this->jwtDecoder = $jwtDecoder ?? new Firebase();
+
+        $this->loginModel = model(TokenLoginModel::class); // @phpstan-ignore-line
+    }
+
+    /**
+     * Attempts to authenticate a user with the given $credentials.
+     * Logs the user in with a successful check.
+     *
+     * @param array{token?: string} $credentials
+     */
+    public function attempt(array $credentials): Result
+    {
+        /** @var IncomingRequest $request */
+        $request = service('request');
+
+        $ipAddress = $request->getIPAddress();
+        $userAgent = $request->getUserAgent();
+
+        $result = $this->check($credentials);
+
+        if (! $result->isOK()) {
```

Comment:
With JWT being stateless, this is going to record a login attempt for every API request. And it looks like it's recording it in the main login table. I think this should be handled like we handled Access Tokens, which at the very least used a different table. 

Now that I'm thinking through this more, I don't know that access tokens or JWT really need to be tracked every page view anyway. Tracking login attempts is more for catching bad actors attempting to break into accounts. With PAT and JWT, I don't believe you typically see these same types of attacks, so rapidly expanding a database to very large sizes doesn't seem beneficial.

Reply:
First, the log is recorded in `auth_token_logins` table from the beginning. Because `tokenLoginModel` is used.

Now I added config for logging. The default is *Record only failures*.